### PR TITLE
introduce concrete Field type for Manifolds

### LINKED ISSUE #62: Use concrete types for the field a manifold is build upon
The additional type `T` we use to distinguish real- and complex-valued manifolds should be a concrete type, for example following https://github.com/Jutho/TensorKit.jl/blob/ae270985e1d71bc982d912144d98ef8b560da9e2/src/spaces/vectorspaces.jl#L10-L19 as suggested here https://github.com/JuliaNLSolvers/Manifolds.jl/pull/49#discussion_r358119733
and here https://github.com/JuliaNLSolvers/Manifolds.jl/pull/49#discussion_r358122101

### PR DESCRIPTION
**Pull request by kellertuer**
to close #62 .

### CODE REVIEW

File: src/Manifolds.jl [hunk 1]
```diff
@@ -104,6 +104,22 @@ using FiniteDifferences
...
 
+"""
+    Field
+
+An abstract type to represent the field matrix manifolds are build upon,
+following the idea of [TensorKit](https://github.com/Jutho/TensorKit.jl)
+in order to hace concrete field types to dispatch on. The two most common
+field types are `RealNumbers` (`ℝ` for short) and `ComplexNumbers` (`ℂ`).
+"""
+abstract type Field end
```
Comment (mateuszbaran) [inline src/Manifolds.jl]: Maybe we should call this `AbstractField`?

File: test/euclidean.jl [hunk 2]
```diff
@@ -2,14 +2,16 @@ include("utils.jl")
...
+    Ec = Manifolds.Euclidean(3;field=ℂ)
     EM = Manifolds.MetricManifold(E,Manifolds.EuclideanMetric())
     @test is_default_metric(EM) == Val{true}()
     @test is_default_metric(E,Manifolds.EuclideanMetric()) == Val{true}()
     x = zeros(3)
     @test det_local_metric(EM,x) == one(eltype(x))
     @test log_local_metric_density(EM,x) == zero(eltype(x))
     @test project_point!(E,x) == x
-
+    @test manifold_dimension(Ec) == 2*manifold_dimension(E)
```
Comment (mateuszbaran) [inline test/euclidean.jl]: I think we should run full tests on `Ec`.

File: src/Manifolds.jl (hunk 1 above)
Comment (kellertuer) [reply src/Manifolds.jl]: I don't mind and it would be consistent with our naming scheme, you're right.

File: test/euclidean.jl (hunk 2 above)
Comment (kellertuer) [reply test/euclidean.jl]: No problem, I can do a few numbers for that manifold (though Euclidean really has a lot of tests already currently).

Comment (mateuszbaran) [reply test/euclidean.jl]: Calling `test_manifold` should be enough I think.

**Comment by mateuszbaran** [comment]
I've added some more tests for the complex Euclidean case. It looks like there were problems in the test suite and random points and tangent vectors don't work right now in the complex case.

**Comment by kellertuer** [comment]
Code coverage is a little strange, since it says, that `Euclidean` was roduced but the homepage says it's still 100%. Is it ok to merge this?

**Comment by mateuszbaran** [comment]
Coverage sometimes behaves strangely, I think it's still OK to merge this. I don't see any changes that could actually reduce the coverage of `Euclidean`.

**Pull request merged**